src/js/scheduler.ts

```diff
@@ -0,0 +1,93 @@
+import {
+  IN_BROWSER_MAIN_THREAD,
+  IN_NODE,
+  IN_BROWSER_WEB_WORKER,
+  IN_SAFARI,
+} from "./environments";
+
+const scheduleCallbackImmediateMessagePrefix =
+  "sched$" + Math.random().toString(36).slice(2) + "$";
+const tasks: Record<number, () => void> = {};
+let nextTaskHandle = 0;
+
+/**
+ * Setup global message event listener to handle immediate callbacks
+ */
+function installPostMessageHandler() {
+  if (!IN_BROWSER_MAIN_THREAD) {
+    return;
+  }
+
+  const onGlobalMessage = (event: MessageEvent) => {
+    if (
+      typeof event.data === "string" &&
+      event.data.indexOf(scheduleCallbackImmediateMessagePrefix) === 0
+    ) {
+      const handle = +event.data.slice(
+        scheduleCallbackImmediateMessagePrefix.length,
+      );
+      const task = tasks[handle];
+      if (task) {
+        try {
+          task();
+        } finally {
+          delete tasks[handle];
+        }
+      }
+    }
+  };
+
+  globalThis.addEventListener("message", onGlobalMessage, false);
+}
+
+installPostMessageHandler();
+
+/**
+ * Implementation of zero-delay scheduler for immediate callbacks
+ * Try our best to use the fastest method available, based on the current environment.
+ * This implementation is based on the following references:
+ *   - https://github.com/YuzuJS/setImmediate
+ *   - https://github.com/zloirock/core-js/blob/master/packages/core-js/internals/task.js
+ * General notes:
+ * - Promise.resolve().then() is not a good option. As it uses microtask queue unlike setTimeout which uses macrotask queue
+ *   - Ref: https://github.com/YuzuJS/setImmediate/pull/56
+ * - MessageChannel is faster (2-3x) than postMessage for Chrome, Firefox, and Node.js
+ *   but it has some issues in Safari (slow and messes up the event loop), and Deno.
+ *   - Ref: https://github.com/pyodide/pyodide/pull/4583
+ *   - Ref: https://github.com/zloirock/core-js/issues/624
+ *   - Ref: https://github.com/YuzuJS/setImmediate/issues/80
+ */
+function scheduleCallbackImmediate(callback: () => void) {
+  if (IN_NODE) {
+    // node has setImmediate, let's use it
+    setImmediate(callback);
+  } else if (IN_BROWSER_MAIN_THREAD) {
+    tasks[nextTaskHandle] = callback;
+    globalThis.postMessage(
+      scheduleCallbackImmediateMessagePrefix + nextTaskHandle,
+      "*",
+    );
+    nextTaskHandle++;
+  } else if (IN_BROWSER_WEB_WORKER && !IN_SAFARI) {
+    const channel = new MessageChannel();
+    channel.port1.onmessage = () => callback();
+    channel.port2.postMessage("");
```

Comment:
In the core-js implementation they only make one MessageChannel and share it for all messages. Is it really a good idea to allocate a separate MessageChannel every time we call this?

Reply:
I tried that in [this commit](https://github.com/pyodide/pyodide/pull/4590/commits/b72cf110b8d2fc76d1b59ce60b0c51ade1a3609a), but somehow there was a failure in Node.js. I don't understand why.